flux-kube/plugins/flux-kube.c

```diff
@@ -0,0 +1,352 @@
+        return -1;
+    }
+    
+    if (flux_future_then (f, -1.0f, get_response_continuation, p) < 0) {
+        flux_jobtap_raise_exception (p, FLUX_JOBTAP_CURRENT_JOB,
+                                     "alloc", 0,
+                                     "flux-kube: %s: Failed to "
+                                     "submit jobspec: %s",
+                                     topic,
+                                     strerror (errno));
+        return -1;
+    }
+
+    if (alloc_start (p, id, R) < 0) {
+        flux_jobtap_raise_exception (p, id, "alloc", 0,
+                                     "failed to start allocation");
+        return -1;
+    }
+
+    if (flux_jobtap_job_set_flag (p,
+                                  FLUX_JOBTAP_CURRENT_JOB,
+                                  "alloc-bypass") < 0) {
+        flux_jobtap_raise_exception (p, FLUX_JOBTAP_CURRENT_JOB,
+                                     "alloc", 0,
+                                     "Failed to set alloc-bypass: %s",
+                                     strerror (errno));
+        return -1;
+    }
+
+    return 0;
+}
+
+static int cleanup_cb (flux_plugin_t *p,
+                       const char *topic,
+                       flux_plugin_arg_t *args,
+                       void *arg)
+{
+    const char *jobspec = NULL;
+    flux_future_t *f;
+    int retval = -1;
+
+    /*  Need to get the jobspec and send it to flux-kube daemon 
+     *  for cancellation.
+     */
+    if (!(jobspec = flux_jobtap_job_aux_get (p,
+                                             FLUX_JOBTAP_CURRENT_JOB,
+                                             "flux-kube::jobspec")))
+        return 0;
+
+    if (!(f = flux_rpc_pack (flux_jobtap_get_flux (p), "fluxkube.cancel", 
+                             FLUX_NODEID_ANY, 0,
+                             "{s:s}", "jobspec", jobspec))) { 
+        flux_jobtap_raise_exception (p, FLUX_JOBTAP_CURRENT_JOB,
+                                     "alloc", 0,
+                                     "flux-kube: %s: Failed to send RPC: %s",
+                                     topic,
+                                     strerror (errno));
+        return -1;
+    }
+    
+    if (flux_future_then (f, -1.0f, get_response_continuation, p) < 0) {
+        flux_jobtap_raise_exception (p, FLUX_JOBTAP_CURRENT_JOB,
+                                     "alloc", 0,
+                                     "flux-kube: %s: Failed to "
+                                     "cancel jobspec: %s",
+                                     topic,
+                                     strerror (errno));
+        return -1;
+    }
+
+    /*  If flux-kube::free is set on this job, then this plugin
+     *  sent an "alloc" event, so a "free" event needs to be sent now.
+     */
+    if (flux_jobtap_job_aux_get (p,
+                                 FLUX_JOBTAP_CURRENT_JOB,
+                                 "flux-kube::free")) {
```

Comment:
Similar to the question above, should the `free` event be postponed until `flluxkube.cancel` returns?

Reply:
I'm not sure the order matters in this case. If `cleanup_cb` is called when the walltime is reached, and `fluxkube.cancel` fails this will likely have to be handled manually. The same is true for a user-initiated job cancellation.

Reply:
See above.